File: archive/pulumi-ts/Pr1584/lib/compute.ts
/**
 * compute.ts
 *
 * This module defines compute-related resources including EC2 instances
 * for the secure AWS infrastructure.
 */
import * as pulumi from '@pulumi/pulumi';
import * as aws from '@pulumi/aws';

export function createEc2Instance(
  environment: string,
  subnetId: pulumi.Output<string>,
  securityGroupId: pulumi.Output<string>,
  instanceType: string,
  provider: aws.Provider
): aws.ec2.Instance {
  // Get the latest Amazon Linux 2023 AMI (more secure and up-to-date)
  const amiId = aws.ec2
    .getAmi(
      {
        mostRecent: true,
        owners: ['amazon'],
        filters: [
          {
            name: 'name',
            values: ['al2023-ami-*-x86_64'],
          },
          {
            name: 'virtualization-type',
            values: ['hvm'],
          },
        ],
      },
      { provider }
    )
    .then(ami => ami.id);

  // Create IAM role for SSM access with least privilege
  const ssmRole = new aws.iam.Role(
    `ssm-role-${environment}`,
    {
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Action: 'sts:AssumeRole',
            Effect: 'Allow',
            Principal: {
              Service: 'ec2.amazonaws.com',
            },
            Condition: {
              StringEquals: {
                'aws:RequestedRegion': pulumi.output(provider.region),
              },
            },
          },
        ],
      }),
      tags: {
        Name: `ssm-role-${environment}`,
        Environment: environment,
        ManagedBy: 'Pulumi',
      },
    },
    { provider }
  );

  // Attach SSM managed policy to the role
  new aws.iam.RolePolicyAttachment(
    `ssm-policy-attachment-${environment}`,
    {
      role: ssmRole.name,
      policyArn: 'arn:aws:iam::aws:policy/AmazonSSMManagedInstanceCore',
    },
    { provider }
  );

  // Add CloudWatch agent policy for monitoring
  new aws.iam.RolePolicyAttachment(
    `cloudwatch-agent-policy-${environment}`,
    {
      role: ssmRole.name,
      policyArn: 'arn:aws:iam::aws:policy/CloudWatchAgentServerPolicy',
    },
    { provider }
  );

  // Create instance profile
  const instanceProfile = new aws.iam.InstanceProfile(
    `instance-profile-${environment}`,
    {
      role: ssmRole.name,
      tags: {
        Name: `instance-profile-${environment}`,
        Environment: environment,
        ManagedBy: 'Pulumi',
      },
    },
    { provider }
  );

  // User data script for basic hardening and setup with SSM agent
  const userData = `#!/bin/bash
set -euo pipefail  # Exit on error, undefined vars, pipe failures

# Log all output
exec > >(tee /var/log/user-data.log)
exec 2>&1

# Update system packages
dnf update -y

# Install AWS CLI v2 (Amazon Linux 2023 uses dnf)
dnf install -y awscli2

# Install and start SSM agent (usually pre-installed on Amazon Linux 2023)
dnf install -y amazon-ssm-agent
systemctl enable amazon-ssm-agent
systemctl start amazon-ssm-agent

# Install CloudWatch agent
dnf install -y amazon-cloudwatch-agent

# Basic security hardening
sed -i 's/#PermitRootLogin yes/PermitRootLogin no/' /etc/ssh/sshd_config
sed -i 's/#PasswordAuthentication yes/PasswordAuthentication no/' /etc/ssh/sshd_config
sed -i 's/#PubkeyAuthentication yes/PubkeyAuthentication yes/' /etc/ssh/sshd_config
sed -i 's/#MaxAuthTries 6/MaxAuthTries 3/' /etc/ssh/sshd_config
sed -i 's/#ClientAliveInterval 0/ClientAliveInterval 300/' /etc/ssh/sshd_config
sed -i 's/#ClientAliveCountMax 3/ClientAliveCountMax 2/' /etc/ssh/sshd_config

# Disable unused services
systemctl disable postfix || true
systemctl stop postfix || true

# Configure automatic security updates
dnf install -y dnf-automatic
sed -i 's/apply_updates = no/apply_updates = yes/' /etc/dnf/automatic.conf
sed -i 's/upgrade_type = default/upgrade_type = security/' /etc/dnf/automatic.conf
systemctl enable dnf-automatic.timer
systemctl start dnf-automatic.timer

# Set up fail2ban for SSH protection
dnf install -y fail2ban
systemctl enable fail2ban
systemctl start fail2ban

# Restart SSH with new configuration
systemctl restart sshd

# Create a non-root user with limited privileges
useradd -m -s /bin/bash -G wheel trainer
# Set up SSH directory for the user
mkdir -p /home/trainer/.ssh
chmod 700 /home/trainer/.ssh
chown trainer:trainer /home/trainer/.ssh

# Set proper file permissions
chmod 700 /home/trainer
chown trainer:trainer /home/trainer

# Configure sudo to require password
echo '%wheel ALL=(ALL) ALL' > /etc/sudoers.d/wheel

echo "Setup completed with SSM agent and security hardening" >> /var/log/user-data.log
`;

  // Create EC2 instance
  const instance = new aws.ec2.Instance(
    `ec2-${environment}`,
    {
      ami: amiId,
      instanceType: instanceType,
      iamInstanceProfile: instanceProfile.name,
      subnetId: subnetId,
      vpcSecurityGroupIds: [securityGroupId],

      userData: userData,

      // Enable detailed monitoring
      monitoring: true,

      // EBS optimization for better performance
      ebsOptimized: true,

      // Enforce IMDSv2 for better security
      metadataOptions: {
        httpEndpoint: 'enabled',
        httpTokens: 'required', // Enforce IMDSv2
        httpPutResponseHopLimit: 1,
        instanceMetadataTags: 'enabled',
      },

      // Root block device configuration
      rootBlockDevice: {
        volumeType: 'gp3',
        volumeSize: 30,
        encrypted: true,
        deleteOnTermination: true,
        tags: {
          Name: `ebs-root-${environment}`,
          Environment: environment,
          ManagedBy: 'Pulumi',
        },
      },

      tags: {
        Name: `ec2-${environment}`,
        Environment: environment,
        Purpose: 'WebServer',
        ManagedBy: 'Pulumi',
        Backup: 'Required',
      },

      // Enable termination protection for production
      disableApiTermination: environment === 'prod',
    },
    {
      provider,
      // Ensure instance is created after instance profile
      dependsOn: [instanceProfile],
    }
  );

  return instance;
}

File: archive/pulumi-py/Pr5996/lib/infrastructure/compute.py
"""
Compute infrastructure module.

This module creates EC2 instances with proper configuration.

"""

import pulumi
import pulumi_aws as aws
from pulumi import Output


class ComputeStack:
    """
    Compute stack that creates EC2 instances.
    
    Creates:
    - EC2 instances across multiple AZs
    - Dynamic AMI lookup
    - Proper user data for monitoring and SSM
    """
    
    def __init__(self, config, provider_manager, private_subnet_ids, 
                 security_group_id, instance_profile_name, instance_profile, parent=None):
        """
        Initialize the compute stack.
        
        Args:
            config: InfraConfig instance
            provider_manager: AWSProviderManager instance
            private_subnet_ids: List of private subnet IDs
            security_group_id: Security group ID
            instance_profile_name: Instance profile name
            instance_profile: Instance profile resource for dependency
            parent: Optional parent resource
        """
        self.config = config
        self.provider_manager = provider_manager
        self.private_subnet_ids = private_subnet_ids
        self.security_group_id = security_group_id
        self.instance_profile_name = instance_profile_name
        self.instance_profile = instance_profile
        self.parent = parent
        
        # Get latest Amazon Linux 2023 AMI dynamically
        self.ami = self._get_latest_ami()
        
        # Create EC2 instances
        self.instances = self._create_ec2_instances()
    
    def _get_latest_ami(self) -> aws.ec2.AwaitableGetAmiResult:
        """
        Get latest Amazon Linux 2023 AMI dynamically.
        
        Fixes hardcoded AMI failure by using dynamic lookup.
        
        Returns:
            AMI data
        """
        ami = aws.ec2.get_ami(
            most_recent=True,
            owners=['amazon'],
            filters=[
                aws.ec2.GetAmiFilterArgs(
                    name='name',
                    values=['al2023-ami-*-x86_64']
                ),
                aws.ec2.GetAmiFilterArgs(
                    name='virtualization-type',
                    values=['hvm']
                ),
                aws.ec2.GetAmiFilterArgs(
                    name='root-device-type',
                    values=['ebs']
                )
            ],
            opts=pulumi.InvokeOptions(provider=self.provider_manager.get_provider())
        )
        
        return ami
    
    def _get_user_data(self, instance_index: int) -> str:
        """
        Generate user data script for EC2 instance.
          
        Args:
            instance_index: Index of the instance
            
        Returns:
            User data script
        """
        user_data = f"""#!/bin/bash
set -e

# Configure logging
exec > >(tee -a /var/log/user-data.log) 2>&1
echo "Starting user data script at $(date)"
echo "Instance: {self.config.project_name}-ec2-{instance_index}-{self.config.environment_suffix}"

# Update system packages
dnf update -y

# Install CloudWatch agent
dnf install -y amazon-cloudwatch-agent

# Install and configure SSM agent (should be pre-installed on AL2023, but ensure it's there)
dnf install -y amazon-ssm-agent
systemctl enable amazon-ssm-agent
systemctl start amazon-ssm-agent
echo "SSM agent status:"
systemctl status amazon-ssm-agent --no-pager

# Create instance info file
cat > /home/ec2-user/instance-info.txt <<'EOFINFO'
Instance: {self.config.project_name}-ec2-{instance_index}-{self.config.environment_suffix}
Region: {self.config.primary_region}
Environment: {self.config.environment_suffix}
Initialized: $(date)
EOFINFO

# Create helper script for S3 interaction
cat > /home/ec2-user/s3-helper.sh <<'EOFSCRIPT'
#!/bin/bash
# Helper script for interacting with S3 buckets
# Usage examples:
#   ./s3-helper.sh upload-data myfile.txt
#   ./s3-helper.sh upload-log application.log
#   ./s3-helper.sh list-data
#   ./s3-helper.sh list-logs

DATA_BUCKET="{self.config.get_resource_name('data-bucket')}"
LOGS_BUCKET="{self.config.get_resource_name('logs-bucket')}"

case "$1" in
    upload-data)
        aws s3 cp "$2" "s3://$DATA_BUCKET/" --sse aws:kms
        echo "Uploaded $2 to $DATA_BUCKET"
        ;;
    upload-log)
        aws s3 cp "$2" "s3://$LOGS_BUCKET/" --sse aws:kms
        echo "Uploaded $2 to $LOGS_BUCKET"
        ;;
    list-data)
        aws s3 ls "s3://$DATA_BUCKET/"
        ;;
    list-logs)
        aws s3 ls "s3://$LOGS_BUCKET/"
        ;;
    *)
        echo "Usage: $0 {{upload-data|upload-log|list-data|list-logs}} [file]"
        exit 1
        ;;
esac
EOFSCRIPT

chmod +x /home/ec2-user/s3-helper.sh
chown ec2-user:ec2-user /home/ec2-user/s3-helper.sh

# Create a test file to verify S3 access
echo "Test file created at $(date)" > /home/ec2-user/test-file.txt
chown ec2-user:ec2-user /home/ec2-user/test-file.txt

echo "User data script completed successfully at $(date)"
"""
        return user_data
    
    def _create_ec2_instances(self) -> list:
        """
        Create EC2 instances across multiple AZs.
        
        Returns:
            List of EC2 instance resources
        """
        instances = []
        
        for i, subnet_id in enumerate(self.private_subnet_ids):
            instance = aws.ec2.Instance(
                f'ec2-instance-{i}',
                ami=self.ami.id,
                instance_type=self.config.instance_type,
                subnet_id=subnet_id,
                vpc_security_group_ids=[self.security_group_id],
                iam_instance_profile=self.instance_profile_name,
                user_data=self._get_user_data(i),
                tags=self.config.get_tags_for_resource(
                    'EC2Instance',
                    Name=self.config.get_resource_name('ec2', str(i))
                ),
                opts=self.provider_manager.get_resource_options(
                    depends_on=[self.instance_profile],
                    parent=self.parent
                )
            )
            
            instances.append(instance)
        
        return instances
    
    def get_instance_ids(self) -> list:
        """Get list of EC2 instance IDs."""
        return [instance.id for instance in self.instances]
    
    def get_instances(self) -> list:
        """Get list of EC2 instance resources."""
        return self.instances

